Block Editor: Consistently render IgnoreNestedEvents for block appender

aduth opened the pull request:
Fixes #18928 
Related: #18379 

This pull request seeks to consistently apply the behavior of `IgnoreNestedEvents` for the block appender component. Previously, the wrapper would have applied to the rendering of the default block appender, and was intended to prevent conflicted interactions between the appender and an ancestor block's focus handlers. Since this was only applied to the default appender and not custom or button appenders, those same conflicts could occur in the latter cases (see #18329). With these changes, `IgnoreNestedEvents` is now used to wrap each possible rendering of the block appender. This also serves as a simplification of the component to unify the render logic.

This also updates the associated `writing-flow.test.js` end-to-end test. Previously, there were assertions which were made to try to verify the intended behavior. However, those assertions relied on the DOM as the source of truth. For the scenario described in #18928, the conflicting focus behavior was such that the active DOM element would be the correct node (the column), but the UI would reflect the incorrect block toolbar (the paragraph), because the selected block in block editor state was prevented from being updated. As of these changes, assertions are made against both the DOM and block editor state to verify correctness.

**Testing Instructions:**

Repeat steps to reproduce from #18928, verifying that you can <kbd>ArrowUp</kbd> from the first block within a column to navigate to the parent Column block.

As described in https://github.com/WordPress/gutenberg/pull/18379#issuecomment-565366937, verify that in a Cover block, you can click the cover block background to select the parent block.

Verify that the end-to-end test included as part of #18379 still passes:

```
npm run test-e2e packages/e2e-tests/specs/editor/blocks/navigation.test.js
```

Verify that the enhanced end-to-end test for writing flow still passes:

```
npm run test-e2e packages/e2e-tests/specs/editor/various/writing-flow.test.js
```

aduth committed:
Block Editor: Consistently render IgnoreNestedEvents for block appender

aduth committed:
Block Editor: Fix grammar in BlockListAppender comment

ZebulanStanphill commented:
I tried out the branch and everything seems to be working! :smile::+1:

tellthemachines commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -290,7 +290,7 @@ function BlockListBlock( {
 	 */
 	const onFocus = () => {
-		if ( ! isSelected && ! isAncestorOfSelectedBlock && ! isPartOfMultiSelection ) {
+		if ( ! isSelected && ! isPartOfMultiSelection ) {
```
I know this is a crucial part to fixing the bug 😅 , but this change has caused the original nav item appender bug to resurface in Firefox (though not in Chrome):
![ezgif com-optimize](https://user-images.githubusercontent.com/8096000/70881344-4c567c00-2020-11ea-8404-1e4458a543d8.gif)
cc @draganescu

draganescu replied:
Yes, Firefox is the issue here and why the whole original attempt was so convoluted :D

aduth replied:
Hm, reading this and #18379, I might expect this could be related a Firefox+macOS specific treatment of click events in buttons in which the clicks do not cause focus:

https://gist.github.com/cvrebert/68659d0333a578d75372
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus

aduth replied:
There's a few related handlings of this in Gutenberg elsewhere:

https://github.com/WordPress/gutenberg/blob/74269e6956392f66622af458b1bb7e8ab1dac3fb/packages/block-editor/src/components/block-list/insertion-point.js#L54-L60

https://github.com/WordPress/gutenberg/blob/74269e6956392f66622af458b1bb7e8ab1dac3fb/packages/block-editor/src/components/default-block-appender/index.js#L38-L49

https://github.com/WordPress/gutenberg/blob/74269e6956392f66622af458b1bb7e8ab1dac3fb/packages/components/src/higher-order/with-focus-outside/index.js#L100-L121

Similarly, I'm thinking it could be enough to add a `tabIndex={ -1 }` to the `div` wrapper, which could allow the focus event to occur when the appender is clicked, while not adding an extra tab stop.

aduth replied:
> Similarly, I'm thinking it could be enough to add a `tabIndex={ -1 }` to the `div` wrapper, which could allow the focus event to occur when the appender is clicked, while not adding an extra tab stop.

This seems to work in my testing. See 9091e9701, which includes the test and an explanatory inline comment.

tellthemachines replied:
It's working fine for me now 🎉

tellthemachines commented on packages/e2e-tests/specs/editor/various/writing-flow.test.js:
```diff
@@ -54,13 +60,19 @@ describe( 'Writing Flow', () => {
 		// Arrow up into nested context focuses last text input
 		await page.keyboard.press( 'ArrowUp' );
+		activeBlockName = await getActiveBlockName();
+		expect( activeBlockName ).toBe( 'core/paragraph' );
```
I wonder if it wouldn't be best to verify instead that the block toolbar is visible at this point. From a user's perspective, that's what really matters, whereas the block being set as active in our data store is more of an implementation detail.

aduth replied:
> I wonder if it wouldn't be best to verify instead that the block toolbar is visible at this point. From a user's perspective, that's what really matters, whereas the block being set as active in our data store is more of an implementation detail.

That's a fair expectation we should want in these tests, I agree. The selected block name was more easily accessible and generally representative of what would be expected to be shown in the UI for the toolbar, but I'll see about making the assertion more targeted to the toolbar.

aduth replied:
In closer inspection, I don't think there's anything of the toolbar markup which can serve as a sufficient indication that the toolbar is showing for a paragraph block:

![image](https://user-images.githubusercontent.com/1779930/70914169-78084080-1fe5-11ea-96b2-93b8136a22d7.png)

I expect this is in-part due to the fact that core block icons are hard-coded, and not part of some keyed collection against which we can reference:

https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/paragraph/icon.js

The only options I could see here is either adding more markup to the toolbar for the sake of the tests, or comparing against this hard-coded SVG value.

For the purposes of this test, I'm inclined to think we can be content with the current assertion against block editor state. If it proves to fail us in the future, it could be worth revisiting.

tellthemachines replied:
Yup, we'll have to revisit if `activeBlockName` at some point gets divorced from toolbar visibility. Might be worth leaving a comment in the test, to make it clear the expectation is that the toolbar is visible at that point.

aduth committed:
Block Editor: Capture button click focus in BlockListAppender

Firefox, Safari compatibility

aduth committed:
Block Editor: Remove redundant IgnoreNestedEvents from InnerBlocks appender

Handled now via ancestor BlockListAppender in which this component would be rendered

tellthemachines reviewed (approved):
Looks good!

draganescu reviewed (approved):
I have tested this as well in both FF and Chrome for the problems which the previous PR tried to solve:

- one click appending of navigation item
- not deselecting a block when the flow block appender was clicked
- maintain one click append for the group block
- fix the new arrow navigation issues and parent block click selection problem whcih the previous PR introduced.

Everything worked great!
